extensions/amp-moviads-player/0.1/amp-moviads-player.js

```diff
@@ -0,0 +1,890 @@
+      const childElements = toArray(sourceElements).concat(
+        toArray(trackElements)
+      );
+      if (childElements.length > 0) {
+        const children = [];
+        childElements.forEach(child => {
+          // Save the first source and first track to preconnect.
+          if (child.tagName == 'SOURCE' && !this.preconnectSource_) {
+            this.preconnectSource_ = child.src;
+          } else if (child.tagName == 'TRACK' && !this.preconnectTrack_) {
+            this.preconnectTrack_ = child.src;
+          }
+          children.push(child./*OK*/ outerHTML);
+        });
+        _this.element.setAttribute(
+          'data-child-elements',
+          JSON.stringify(children)
+        );
+      }
+      // Handle IMASetting JSON
+      const scriptElement = childElementsByTag(_this.element, 'SCRIPT')[0];
+      if (scriptElement && isJsonScriptTag(scriptElement)) {
+        _this.element.setAttribute(
+          'data-ima-settings',
+          scriptElement./*OK*/ innerHTML
+        );
+      }
+      _this.loadMethod();
+      _this.mute();
+    };
+  }
+
+  /** @override */
+  preconnectAds() {
+    const {element, preconnect} = this;
+    if (typeof preconnect !== 'undefined') {
+      preconnect.preload(
+        'https://imasdk.googleapis.com/js/sdkloader/ima3.js',
+        'script'
+      );
+      const source = this.configLocal_.vastUrl;
+      if (source) {
+        preconnect.url(source);
+      }
+      if (this.preconnectSource_) {
+        preconnect.url(this.preconnectSource_);
+      }
+      if (this.preconnectTrack_) {
+        preconnect.url(this.preconnectTrack_);
+      }
+
+      preloadBootstrap(this.win, preconnect);
+    }
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  getConsentPolicy() {
+    return null;
+  }
+
+  /** @override */
+  layoutLoad() {
+    const {element, win} = this;
+    const consentPolicyId = super.getConsentPolicy();
+    const consentPromise = consentPolicyId
+      ? getConsentPolicyState(element, consentPolicyId)
+      : Promise.resolve(null);
+
+    element.setAttribute('data-tag', this.configLocal_.vastUrl);
+    element.setAttribute(
+      'data-poster',
+      this.recommendedPoster_ !== ''
+        ? this.recommendedPoster_
+        : this.config_.moviePosterSrc
+    );
+    element.setAttribute(
+      'data-src',
+      this.recommendedSrc_ !== '' ? this.recommendedSrc_ : this.config_.dataSrc
+    );
+    element.setAttribute(
+      'data-ad-label',
+      this.config_['langText']['advertisement'] + ' (%s / %s)'
+    );
+
+    this.resizeContainerPlayer();
+
+    if (true === this.config_.autoplay) {
+      element.setAttribute('autoplay', true);
+    }
+
+    return consentPromise.then(initialConsentState => {
+      const iframe = getIframe(
+        win,
+        element,
+        'ima-video',
+        {initialConsentState},
+        {allowFullscreen: true}
+      );
```

Comment:
If this is `ima-video`, why can't you just use `amp-ima-video`?

Reply:
@kkloc-smi

Reply:
@alanorozco 

The Player is an integral part of the Video platform that supports large publishers. Without downloading the configuration from an external server, it is impossible to manage ad emissions on websites

Reply:
> downloading the configuration from an external server

This can be done on an intermediate page using `amp-video-iframe` in that case. I'm still not sure why we would need anything custom on the AMP host's layer.

Reply:
@alanorozco 

We cannot use amp-video-iframe because the extension does not work properly in domains other than the original one, and google creates a copy of the page and displays it to users in its domain. Our player is then blocked by browsers. Users then cannot access the video content. This situation is problematic for the website owner and its users.

Example:

https://www.national-geographic.pl/artykul/amp/giewont-zdobyty-na-krzyzu-w-tatrach-zawisl-baner-strajku-kobiet

https://www.google.com/amp/s/www.national-geographic.pl/artykul/amp/od-grenkieta-oderwal-sie-ogromny-lodowiec-ma-wielkosc-paryza#developer

The amp-video-iframe extension is not optimized for video visibility. Our player has mechanisms that stop the playback of video material when less than 50% of the video material is visible in the visible area.

This is in line with the IAB guidelines. Which we want to implement because we want to offer publishers a solution that increases the user experience of their websites.

Our solution enables remote control of video content configuration from the level of our platform (autoplay, advertising server, sound mute), which we offer to our partners. Thanks to this, publishers can manage video materials on their websites from one place. Thanks to this solution, the editorial offices of publishing houses cooperating with us will obtain a solution that will significantly improve the quality of their work and improve the comfort of users using their websites. The amp-video-iframe extension does not offer this functionality.

Reply:
> the extension does not work properly in domains other than the original one

amp-video-iframe **should** work fine from any AMP cache. Are you experiencing this with any amp-video-iframe or just with your own document.

> This is in line with the IAB guidelines. Which we want to implement because we want to offer publishers a solution that increases the user experience of their websites.

`autoplay` on AMP follows this behavior. If not autoplay, you can also get intersection visibility from amp-video-iframe.

> Our solution enables remote control of video content configuration from the level of our platform (autoplay, advertising server, sound mute)

1. `autoplay` and `mute` should **not** be independent of the autoplay attribute (which you get from amp-video-iframe). A player extension **cannot** be submitted if it autoplays without the attribute or if it doesn't respect `mute`

2. You can still control the advertising because it should be rendered inside the frame.

Reply:
Reminding also that the I2I linked has **not** been approved nor has it gone through design review, which is a requirement.